File: src/main/java/io/crm/core/Reader.java
package io.crm.core;

import io.crm.core.noop.NoopArg;
import io.smallrye.mutiny.Uni;

/**
 * Created by xiongxl in 2022/3/9
 * 值读取器，返回带有值的Future，如果带的值null表示指定的键对应的值不存在
 * @param <K> 键类型
 * @param <V> 值类型
 */
@FunctionalInterface
public interface Reader<K, V> {
    /**
     * 读取操作
     * @param key 指定键
     * @return 返回带有值的Future，如果带的值null表示指定的键对应的值不存在
     */
    Uni<V> read(K key);

    static <K, V> Reader1<K, V, NoopArg> toReader1(Reader<K, V> reader) {
        if (reader == null) {
            return null;
        }
        return (key, arg) -> reader.read(key);
    }
}

File: src/main/kotlin/io/crm/core/kotlin/Reader.kt
package io.crm.core.kotlin

import io.crm.core.Reader
import io.crm.core.Reader1
import io.crm.core.noop.NoopArg

/**
 * Created by xiongxl in 2022/4/9 可挂起版[io.crm.core.Reader]
 */
fun interface Reader<K, V> {
    suspend fun read(key: K): V?

    companion object {
        fun <K, V> toReader1(reader: Reader<K, V>?): Reader1<K, V, NoopArg>? {
            return if (reader == null) {
                null
            } else Reader1 { key: K, _: NoopArg -> reader.read(key) }
        }
    }
}